All project code/index.js

```diff
@@ -79,32 +86,58 @@ app.get("/login", (req, res) => {
 });
 
 // Handles user login from a /login POST request.
-app.post("/login", (req, res) => {
+app.post("/login", async (req, res) => {
     const username = req.body.username;
     const password = req.body.password;
-
     const query = "SELECT * FROM users WHERE username = $1";
 
-    db.any(query, [username])
-    .then(async (user) => {
-        const match = await bcrypt.compare(password, user[0].password);
-
-        if (match)
-        {
-            req.session.user = "TEST";
-            req.session.save();
-            res.redirect("/main");
-        }
-        else
-        {
-            console.log("Incorrect username or password.");
-            res.redirect("/login");
-        }
+    let user; // The user found (or not found) in the database
+
+    // Queries database to find matching user for the inputted username.
+    await db.any(query, [username])
+    .then((foundUser) => {
+        user = foundUser[0];
     })
     .catch((err) => {
-        console.log(err.message);
+        // Something went wrong, refresh the page for the user to try again.
         res.redirect("/login");
+        return console.log(err.message);
     });
+
+    const incorrect_login_feedback = "Incorrect username or password."
+
+    if (user == null) {
+        // User not found in database, so username does not exist.
+        console.log(incorrect_login_feedback);
+        res.render("pages/login", {
```

Comment:
Is there a reason you are handling this outside of the async call to the database? It would make for cleaner code than setting user inside the .then() and much less prone to error because everything would be caught by the catch() statement instead. Usually if some code is dependent on an async call you want to wrap all of that inside of the callback function instead of all of these tests outside of the loop

Reply:
That could work, but in the event that something goes wrong with the request itself the user wouldn't receive different feedback from incorrect username/password.

Reply:
I tried doing it the way you suggested, but it may cause some bad race conditions. Also, this solution would crash the server whenever the username was wrong, because the user == null check never actually happened. A valid, successful response is present when the username is incorrect (so it doesn't hit the catch()) but with no data, causing the server to crash when I ask for data within user (specifically, user.password). I could do a null check in this case, but at that point we're back to my current solution.